Marbles Forever sample asset

echadwick-artist opened the pull request:
Adding this work-in-progress physics sample asset. The asset needs further editing, but it is at a point where others could potentially help improve it. See the readme for details.

![screenshot_Large](https://github.com/user-attachments/assets/7a0dd071-e33f-43f3-8464-8074c0baf19b)

echadwick-artist committed:
Marbles Forever sample asset

Adding this work-in-progress physics sample asset. The asset need further editing, but it is at a point where others could potentially help improve it.

echadwick-artist committed:
spelling fix

I had misspelled Eoin's last name, sorry Eoin!

cx20 commented:
I tried this model in the Babylon.js Playground.
Sorry if I'm using it incorrectly.
It seems to be rotating in a strange direction.

https://playground.babylonjs.com/#8L313O#7
![babylonjs_gltf_physics_animation_001](https://github.com/user-attachments/assets/79239251-3255-4919-bd2e-c0d5dda5a29c)

echadwick-dgg3d commented:
Yes, it will only work in a glTF viewer that supports the current state of these new physics extensions. Instructions for this are in the readme. As far as I know, only Eoin's fork supports this, see [glTF_Physics_Babylon Viewer](https://eoineoineoin.github.io/glTF_Physics_Babylon/packages/demo/dist/#sceneIndex=0).

echadwick-artist committed:
added extensionsRequired

  "extensionsRequired": [
    "KHR_physics_rigid_bodies",
    "KHR_implicit_shapes"
  ],

echadwick-artist commented:
Thank you cx20! I am a big fan of your work, and I know your work is helpful to the whole glTF community!

echadwick-artist committed:
Asset updates

Most of the physics issues have been resolved, and duplicate textures have been removed.

Special thanks to @eoineoineoin for help with the physics settings!

echadwick-artist committed:
Updates

* Archimedes screw added, one alternative for moving the balls up the elevator.
* Trigger volumes added, another alternative for the elevator, if integrated with the Interactivity extension.
* Balls have been set to Combine Mode: Minimum in an attempt to resolve some physics issues.

echadwick-artist committed:
Screw fixed, collision shape adjustments

* extensionsRequired was fixed to require the physics extensions.
* Combine for the balls was set to Maximum. Can't see the difference between the modes though, just shooting in the dark here.
* Archimedes screw now rotates properly: unparented from the root Compound parent, set to Infinite Mass, and Angular Velocity was set to negative.
* Vertex positions were edited on physics shapes at top and bottom of elevator shaft, to prevent the balls from getting stuck there.
* Bottom gravity trigger for Interactivity extension was repositioned so ball gets to the center of the elevator shaft before it enters the trigger.
* Tried enabling tangents on export, but that caused more validation problems than without it, because the exporter seems to add tangents to ALL meshes, regardless of whether they have a normal map or not. So, tangent export was deactivated.

echadwick-artist commented:
![2026-03-13_issues](https://github.com/user-attachments/assets/c5d24c87-f2bf-42a9-9b1a-b45dd72e0825)

A few problems remain and I don't know how to fix these:
1. Elevator emissive UV animation is not playing in the Babylon viewer.
2. Ball materials are looking very odd in glTF Sample Viewer.
3. Balls are moving very very slowly in glTF Sample Viewer.

The two renderers I tested with:
* [glTF_Physics_Babylon Viewer](https://eoineoineoin.github.io/glTF_Physics_Babylon/packages/demo/dist/#sceneIndex=0).
* [glTF Sample Viewer with Physics](https://github.khronos.org/glTF-Sample-Viewer-Release/physics/)

echadwick-artist committed:
Compound Parents are non-renderable

* Re-exported with v0.6.1 of Blender physics exporter (https://github.com/eoineoineoin/glTF_Physics_Blender_Exporter/releases)
* Compound Parent meshes are now properly not rendered in both Babylonjs Physics Viewer and glTF Sample Viewer Physics.

echadwick-dgg3d committed:
Updated documentation

Updated the making-of information.
Added explanatory images.
Updated the "issues" image.
